unsloth/mlx/lora.py

```diff
@@ -0,0 +1,105 @@
+from pathlib import Path
+import math
+import mlx.nn as nn
+import mlx.optimizers as optim
+import mlx.core as mx
+from .trainer.trainer import TrainingArgs, TrainingCallback, train
+from .trainer.utils import (
+    build_schedule,
+    linear_to_lora_layers,
+    print_trainable_parameters,
+)
+from .mlx_utils import  save_config
+
+def train_model(
+    args,
+    model: nn.Module,
+    tokenizer,
+    train_set,
+    valid_set,
+    training_callback: TrainingCallback = None,
+):
+    model.freeze()
+    linear_to_lora_layers(
+        model,
+        min(args.r,len(model.layers)/2),
+        {"rank": args.r, "alpha": args.lora_alpha, "dropout": args.lora_dropout, "scale": float(args.lora_alpha)/math.sqrt(float(args.r)) if args.use_rslora else float(args.lora_alpha)/float(args.r)},
+    )
+    print_trainable_parameters(model)
+
+    adapter_path = Path(args.save_path)
+    adapter_path.mkdir(parents=True, exist_ok=True)
+
+    adapter_file = adapter_path / args.adapter_file
+    config = {
+        "num_layers" : min(args.r,len(model.layers)/2),
+        "lora_parameters" : {"rank": args.r, "alpha": args.lora_alpha, "dropout": args.lora_dropout, "scale": float(args.lora_alpha)/math.sqrt(float(args.r)) if args.use_rslora else float(args.lora_alpha)/float(args.r)}
+    }
+    save_config(config, adapter_path / "adapter_config.json")
+
+    # init training args
+    training_args = TrainingArgs(
+        batch_size=args.per_device_train_batch_size,
+        iters=args.max_steps,
+        val_batches=25,
+        steps_per_report=args.logging_steps,
+        steps_per_eval=200,
+        steps_per_save=100,
+        adapter_file=adapter_file,
+        max_seq_length=args.max_seq_length,
+        grad_checkpoint=args.use_gradient_checkpointing,
+    )
+
+    mx.random.seed(args.seed)
+    model.train()
+    if args.lr_scheduler_type == "linear":
+        arguments = [0.0,args.learning_rate,args.warmup_steps]
```

Comment:
```suggestion
        arguments = [args.learning_rate,0.0,args.max_steps-args.warmup_steps]
```

This will appropriately handle learning rate and decay. The prior configuration sets the start (post warmup at 0.0. It also decayed in steps = warmup instead of the max_steps - warmup (Decay of 5 vs. 95)